File: SortVision/public/code/insertion/csharp/insertionSort.cs
using System;

public static class InsertionSort {
    // Time: Best O(n), Avg/Worst O(n^2), Space: O(1)
    public static void Sort(int[] arr) {
        if (arr == null || arr.Length <= 1) return;

        for (int i = 1; i < arr.Length; i++) {
            int key = arr[i];
            int j = i - 1;
            while (j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j];
                j--;
            }
            arr[j + 1] = key;
        }
    }

    public static void Main() {
        int[] arr = {64, 25, 12, 22, 11};
        Sort(arr);
        Console.WriteLine(string.Join(", ", arr));
    }
}

File: SortVision/public/code/insertion/haskell/insertionSort.hs
module InsertionSort (insertionSort) where

-- Time: Best O(n), Avg/Worst O(n^2), Space: O(n) for immutable list rebuilds
insertionSort :: Ord a => [a] -> [a]
insertionSort = foldl insert []
  where
    insert [] x = [x]
    insert (y:ys) x
      | x <= y = x : y : ys
      | otherwise = y : insert ys x

main :: IO ()
main = print (insertionSort [64, 25, 12, 22, 11])
